pkg/kubectl/cmd/apply_test.go

```diff
@@ -990,6 +998,91 @@ func TestRunApplySetLastApplied(t *testing.T) {
+func TestRunApplyDiffLastApplied(t *testing.T) {
+	initTestErrorHandler(t)
+	rcByte := readBytesFromFile(t, filenameApplyDiffRESP)
+	pathRC := "/namespaces/test/replicationcontrollers/" + "test-rc"
+
+	rcChangedByte := readBytesFromFile(t, filenameApplyDiffChangedRESP)
+
+	tests := []struct {
+		name, nameRC, pathRC, filePath, expectedGitDiffOut, expectedDiffOut, expectedErr, output string
+		resp                                                                                     []byte
+	}{
+		{
+			name:               "git-diff objects with no different",
+			filePath:           filenameRC,
+			expectedErr:        "",
+			expectedGitDiffOut: "",
+			expectedDiffOut:    "",
+			resp:               rcByte,
+			output:             "yaml",
+		},
+		{
+			name:               "git-diff after local file changed with json",
+			filePath:           filenameLocalChangedRC,
+			expectedErr:        "",
+			expectedGitDiffOut: "-                \"containerPort\": 123\n+                \"containerPort\": 80\n               }\n",
+			expectedDiffOut:    "<                 \"containerPort\": 123\n>                 \"containerPort\": 80\n",
+			resp:               rcByte,
+			output:             "json",
+		},
+		{
+			name:               "git-diff after annotations changed",
+			filePath:           filenameRC,
+			expectedErr:        "",
+			expectedGitDiffOut: "-        - containerPort: 80\n+        - containerPort: 123\n",
+			expectedDiffOut:    "<         - containerPort: 80\n>         - containerPort: 123\n",
+			resp:               rcChangedByte,
+			output:             "yaml",
+		},
+	}
+	for _, test := range tests {
+		f, tf, _, _ := cmdtesting.NewAPIFactory()
+		tf.Printer = &testPrinter{}
+		tf.UnstructuredClient = &fake.RESTClient{
+			APIRegistry:          api.Registry,
+			NegotiatedSerializer: unstructuredSerializer,
+			Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+				switch p, m := req.URL.Path, req.Method; {
+				case p == pathRC && m == "GET":
+					bodyRC := ioutil.NopCloser(bytes.NewReader(test.resp))
+					return &http.Response{StatusCode: 200, Header: defaultHeader(), Body: bodyRC}, nil
+				default:
+					t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
+					return nil, nil
+				}
+			}),
+		}
+		tf.Namespace = "test"
+		tf.ClientConfig = defaultClientConfig()
+		buf, errBuf := bytes.NewBuffer([]byte{}), bytes.NewBuffer([]byte{})
+
+		cmd := NewCmdApplyDiffLastApplied(f, buf, errBuf)
+		os.Setenv("KUBECTL_EXTERNAL_DIFF", "testdata/diff/test_diff.sh")
+
+		cmd.Flags().Set("filename", test.filePath)
+		cmd.Flags().Set("output", test.output)
+		cmd.Run(cmd, []string{})
+
+		if pathExist("/tmp/git-diff.result") {
+			out := string(readBytesFromFile(t, "/tmp/git-diff.result"))
+			if out != test.expectedGitDiffOut {
+				t.Fatalf("%s: unexpected output: %s\nexpected: %s", test.name, out, test.expectedGitDiffOut)
+			}
+			os.Remove("/tmp/git-diff.result")
+		}
+
+		if pathExist("/tmp/diff.result") {
```

Comment:
These filenames definitely need to be constants defined at the top of the function.